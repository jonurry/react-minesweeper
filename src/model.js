export const FLAGS = { none: 0, mine: 1, possible: 2 };
export const GAME_STATUS = { initialised: 0, started: 1, won: 2, lost: 3 };

function getNearestNeighbours(position) {
  // work out the nearest neighbours for the given minefield location
  let neighbours = [];
  let columns = this.columns;
  // is the current position on the left of the minefield?
  let left = position % columns === 0;
  // is the current position on the right of the minefield?
  let right = (position + 1) % columns === 0;
  let nw = left ? undefined : this.minefield[position - columns - 1];
  let n = this.minefield[position - columns];
  let ne = right ? undefined : this.minefield[position - columns + 1];
  let w = left ? undefined : this.minefield[position - 1];
  let e = right ? undefined : this.minefield[position + 1];
  let sw = left ? undefined : this.minefield[position + columns - 1];
  let s = this.minefield[position + columns];
  let se = right ? undefined : this.minefield[position + columns + 1];
  // prettier-ignore
  neighbours = [
    nw, n, ne,
     w,     e,
    sw, s, se
  ];
  // exclude neighbours that are outside the minefield
  neighbours = neighbours.filter(value => {
    return value !== undefined;
  });
  return neighbours;
}

const getRandomIntInclusive = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min; // The minimum and maximum are inclusive
};

function placeMinesRandomlyInMinefield() {
  const upperBound = this.spaces - 1;
  for (let i = 0; i < this.mines; i++) {
    let position;
    do {
      position = getRandomIntInclusive(0, upperBound);
    } while (this.minefield[position].value === '*');
    this.minefield[position].value = '*';
  }
}

function revealMines() {
  for (let item of this.minefield) {
    if (item.value === '*') {
      item.revealed = true;
    }
  }
}

class Model {
  constructor(spaces, mines, columns) {
    this.getNearestNeighbours = getNearestNeighbours.bind(this);
    this.placeMinesRandomlyInMinefield = placeMinesRandomlyInMinefield.bind(
      this
    );
    this.revealMines = revealMines.bind(this);
    this.initialiseMinefield(spaces, mines, columns);
  }

  cycleFlag(position) {
    // if the position is invalid return FLAGS.none
    if (position < 0 || position >= this.spaces) {
      return FLAGS.none;
    }
    // cycle through flags
    // the enumeration is in the order that the flags cycle through
    // so adding one will cycle through the flags in the correct order
    // once upper bound is reached reset to beginning
    let flag = this.minefield[position].flag + 1;
    if (flag > FLAGS.possible) {
      flag = FLAGS.none;
    }
    this.minefield[position].flag = flag;
    // return the new flag at the given position
    return flag;
  }

  getContent(position) {
    // if the position is invalid return 0
    if (position < 0 || position >= this.minefield.length) {
      return 0;
    }
    // return the content at the given position
    return this.minefield[position].value;
  }

  getFlag(position) {
    // if the position is invalid return no flag
    if (position < 0 || position >= this.minefield.length) {
      return FLAGS.none;
    }
    // return the flag at the given position
    return this.minefield[position].flag;
  }

  initialiseMinefield(spaces = 72, mines = 10, columns = 8) {
    // number of spaces should be at least 72
    // number of spaces should be exactly divisible by 72
    //   this is so that all rows are full at all screen sizes
    // number of mines should not exceed 50% of the available space

    // round the available space to the nearest factor of 72
    const divisibleBy = 72;
    if (spaces < 72) {
      spaces = 72;
    }
    this.spaces = Math.ceil(spaces / divisibleBy) * divisibleBy;
    const maxMines = this.spaces / 2;

    if (mines > maxMines) {
      this.mines = maxMines;
    } else if (mines < 10) {
      this.mines = 10;
    } else {
      this.mines = mines;
    }
    this.spacesLeftToReveal = this.spaces - this.mines;
    this.minefield = new Array(this.spaces).fill().map((curr, index) => ({
      flag: FLAGS.none,
      id: index,
      revealed: false,
      value: 0
    }));
    this.columns = columns;
    this.gameStatus = GAME_STATUS.initialised;
    this.trippedMineId = -1;
    this.placeMinesRandomlyInMinefield();
    this.populateNumberOfNearestMines();
  }

  isRevealed(position) {
    // if the position is invalid return false
    if (position < 0 || position >= this.minefield.length) {
      return false;
    }
    // return whether the content is revealed or not at the given position
    return this.minefield[position].revealed;
  }

  populateNumberOfNearestMines() {
    // work out the number of nearest mines for each minefield location
    let that = this;
    this.minefield.map((x, i) => {
      if (x.value !== '*') {
        let neighbours = getNearestNeighbours.call(that, i);
        let mines = neighbours.reduce((acc, curr) => {
          if (curr.value === '*') {
            return acc + 1;
          } else {
            return acc;
          }
        }, 0);
        x.value = mines;
      }
      return x;
    });
  }

  reveal(position) {
    let that = this;
    // if the position is invalid, or the game is over, return 0
    if (
      position < 0 ||
      position >= this.minefield.length ||
      this.gameStatus === GAME_STATUS.won ||
      this.gameStatus === GAME_STATUS.lost
    ) {
      return 0;
    }
    if (!this.isRevealed(position)) {
      // mark content as revealed
      this.minefield[position].revealed = true;
      this.spacesLeftToReveal--;
    }
    // store the content of the revealed position
    let content = this.minefield[position].value;
    // Start the game if it isn't already started
    if (this.gameStatus === GAME_STATUS.initialised) {
      this.gameStatus = GAME_STATUS.started;
    }
    // if a mine has been revealed then it's game over
    if (content === '*') {
      this.gameStatus = GAME_STATUS.lost;
      this.trippedMineId = position;
      this.revealMines();
    }
    // if all spaces are revealed except for mines then the game is won
    else if (this.spacesLeftToReveal === 0) {
      this.gameStatus = GAME_STATUS.won;
      this.revealMines();
    } else if (content === 0) {
      // get nearest neighbours to reveal
      let neighbours = getNearestNeighbours.call(that, position);
      // filter out neighbours that have already been revealed
      neighbours = neighbours.filter(value => {
        return !value.revealed;
      });
      // recursively reveal each neighbour
      neighbours.forEach(element => {
        if (!element.revealed) {
          this.reveal(element.id);
        }
      });
    }
    // return the content at the given position
    return content;
  }
}

// iterate over Model using a generator
Model.prototype[Symbol.iterator] = function*() {
  for (let item of this.minefield) {
    yield item;
  }
};

export default Model;
