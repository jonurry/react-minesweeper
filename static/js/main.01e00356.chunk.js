(window.webpackJsonp=window.webpackJsonp||[]).push([[0],{101:function(e,t,i){},103:function(e,t,i){},237:function(e,t,i){"use strict";i.r(t);var n=i(0),a=i.n(n),s=i(14),l=i.n(s),r=(i(98),i(101),i(29)),c=i(21),o=i(22),d=i(40),m=i(37),u=i(39),h=i(28),f=i(27),v=i(87),p=i.n(v),g=i(38),y=i.n(g),E=(i(103),i(88)),M=i.n(E),w=i(89),b=i.n(w),S=i(90),_=i.n(S),C=i(41),L=i.n(C),x=i(91),k=i.n(x),z=i(92),F=i.n(z),T=i(52),N=i.n(T),O=Object(f.createMuiTheme)({palette:{primary:{light:"#9be7ff",main:"#64b5f6",dark:"#2286c3",contrastText:"#263238"},secondary:{light:"#ffe97d",main:"#ffb74d",dark:"#c88719",contrastText:"#263238"},state:{bad:"#e57373",good:"#81c784"}}}),j={margin:"auto"},H=function(e){function t(e){var i;return Object(c.a)(this,t),(i=Object(d.a)(this,Object(m.a)(t).call(this,e))).componentWillUnmount=function(){i.resetTimer()},i.getElapsedTime=function(){var e=(new Date).getTime()-i.startTime,t=Math.floor(e%36e5/6e4),n=Math.floor(e%6e4/1e3);return"".concat(t,":").concat(n.toString().padStart(2,"0"))},i.resetTimer=function(){0!==i.timerId&&clearInterval(i.timerId),i.startTime=null,i.timerId=0,i.setState({time:null})},i.startTimer=function(){i.startTime=(new Date).getTime(),i.timerId=setInterval(function(){i.setState({time:i.getElapsedTime()})},1e3)},i.stopTimer=function(){0!==i.timerId&&clearInterval(i.timerId)},i.classes=e.classes,i.state={time:null},i}return Object(u.a)(t,e),Object(o.a)(t,[{key:"render",value:function(){return a.a.createElement(y.a,{className:this.classes.margin,variant:"headline",color:"inherit"},this.state.time?this.state.time:"0:00")}}]),t}(n.Component),I=i(53),B=i.n(I),D=0,V=1,R=2,W={initialised:0,started:1,won:2,lost:3};function G(e){var t=this.columns,i=e%t===0,n=(e+1)%t===0;return[i?void 0:this.minefield[e-t-1],this.minefield[e-t],n?void 0:this.minefield[e-t+1],i?void 0:this.minefield[e-1],n?void 0:this.minefield[e+1],i?void 0:this.minefield[e+t-1],this.minefield[e+t],n?void 0:this.minefield[e+t+1]].filter(function(e){return void 0!==e})}var A=function(e,t){return e=Math.ceil(e),t=Math.floor(t),Math.floor(Math.random()*(t-e+1))+e};var U=function(){function e(t,i,n){Object(c.a)(this,e),this.getNearestNeighbours=G.bind(this),this.placeMinesRandomlyInMinefield=function(){for(var e=this.spaces-1,t=0;t<this.mines;t++){var i=void 0;do{i=A(0,e)}while("*"===this.getContent(i));this.setContent(i,"*")}}.bind(this),this.relocateMine=function(e){if("*"===this.getContent(e)){var t,i=this.spaces-1;do{t=A(0,i)}while("*"===this.getContent(t));this.setContent(e,0),this.setContent(t,"*"),this.populateNumberOfNearestMines()}}.bind(this),this.revealMines=function(){var e=!0,t=!1,i=void 0;try{for(var n,a=this.minefield[Symbol.iterator]();!(e=(n=a.next()).done);e=!0){var s=n.value;"*"===s.value&&(s.revealed=!0)}}catch(l){t=!0,i=l}finally{try{e||null==a.return||a.return()}finally{if(t)throw i}}}.bind(this),this.initialiseMinefield(t,i,n)}return Object(o.a)(e,[{key:"cycleFlag",value:function(e){if(e<0||e>=this.spaces)return D;var t=this.minefield[e].flag+1;return t>R&&(t=D),this.minefield[e].flag=t,t}},{key:"getContent",value:function(e){return e<0||e>=this.minefield.length?0:this.minefield[e].value}},{key:"getFlag",value:function(e){return e<0||e>=this.minefield.length?D:this.minefield[e].flag}},{key:"initialiseMinefield",value:function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:72,t=arguments.length>1&&void 0!==arguments[1]?arguments[1]:10,i=arguments.length>2&&void 0!==arguments[2]?arguments[2]:8;e<72&&(e=72),this.spaces=72*Math.ceil(e/72);var n=this.spaces/2;this.mines=t>n?n:t<10?10:t,this.spacesLeftToReveal=this.spaces-this.mines,this.minefield=new Array(this.spaces).fill().map(function(e,t){return{delay:0,flag:D,id:t,revealed:!1,value:0}}),this.columns=i,this.gameStatus=W.initialised,this.trippedMineId=-1,this.placeMinesRandomlyInMinefield(),this.populateNumberOfNearestMines()}},{key:"isRevealed",value:function(e){return!(e<0||e>=this.minefield.length)&&this.minefield[e].revealed}},{key:"populateNumberOfNearestMines",value:function(){var e=this;this.minefield.map(function(t,i){if("*"!==t.value){var n=G.call(e,i).reduce(function(e,t){return"*"===t.value?e+1:e},0);t.value=n}return t})}},{key:"reveal",value:function(e){var t=this,i=arguments.length>1&&void 0!==arguments[1]&&arguments[1];if(e<0||e>=this.minefield.length||this.gameStatus===W.won||this.gameStatus===W.lost)return 0;this.gameStatus===W.initialised&&"*"===this.getContent(e)&&this.relocateMine(e),this.isRevealed(e)||(this.minefield[e].revealed=!0,this.spacesLeftToReveal--,i&&(this.minefield[e].delay=A(1,5)));var n=this.getContent(e);if(this.gameStatus===W.initialised&&(this.gameStatus=W.started),"*"===n)this.gameStatus=W.lost,this.trippedMineId=e,this.revealMines();else if(0===this.spacesLeftToReveal)this.gameStatus=W.won,this.revealMines();else if(0===n){var a=G.call(this,e);(a=a.filter(function(e){return!e.revealed})).forEach(function(e){e.revealed||t.reveal(e.id,!0)})}return n}},{key:"setContent",value:function(e,t){e>=0&&e<this.minefield.length&&(this.minefield[e].value=t)}}]),e}();U.prototype[Symbol.iterator]=B.a.mark(function e(){var t,i,n,a,s,l;return B.a.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:t=!0,i=!1,n=void 0,e.prev=3,a=this.minefield[Symbol.iterator]();case 5:if(t=(s=a.next()).done){e.next=12;break}return l=s.value,e.next=9,l;case 9:t=!0,e.next=5;break;case 12:e.next=18;break;case 14:e.prev=14,e.t0=e.catch(3),i=!0,n=e.t0;case 18:e.prev=18,e.prev=19,t||null==a.return||a.return();case 21:if(e.prev=21,!i){e.next=24;break}throw n;case 24:return e.finish(21);case 25:return e.finish(18);case 26:case"end":return e.stop()}},e,this,[[3,14,18,26],[19,,21,25]])});var J=U,P=i(16),Q=i.n(P);function $(e){return a.a.createElement(Q.a,e,a.a.createElement("g",{id:"Bounding_Boxes"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy_3",display:"none"}),a.a.createElement("path",{fill:"none",d:"M0,0h24v24H0V0z"})),a.a.createElement("g",{id:"Duotone"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy_2",display:"none"}),a.a.createElement("g",null,a.a.createElement("polygon",{opacity:"0.3",points:"12.36,6 7,6 7,12 14.24,12 14.64,14 18,14 18,8 12.76,8 \t\t"}),a.a.createElement("path",{d:"M14.4,6L14,4H5v17h2v-7h5.6l0.4,2h7V6H14.4z M18,14h-3.36l-0.4-2H7V6h5.36l0.4,2H18V14z"}))))}function q(e){return a.a.createElement(Q.a,e,a.a.createElement("g",{id:"Bounding_Boxes"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy_3"}),a.a.createElement("path",{fill:"none",d:"M0,0h24v24H0V0z"})),a.a.createElement("g",{id:"Duotone"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy_2"}),a.a.createElement("g",null,a.a.createElement("path",{fillOpacity:"0.3",d:"M18,9.52V6h-3.52L12,3.52L9.52,6H6v3.52L3.52,12L6,14.48V18h3.52L12,20.48L14.48,18H18v-3.52L20.48,12 L18,9.52z M12,18c-3.31,0-6-2.69-6-6s2.69-6,6-6s6,2.69,6,6S15.31,18,12,18z"}),a.a.createElement("path",{d:"M20,8.69V4h-4.69L12,0.69L8.69,4H4v4.69L0.69,12L4,15.31V20h4.69L12,23.31L15.31,20H20v-4.69L23.31,12L20,8.69z M18,14.48 V18h-3.52L12,20.48L9.52,18H6v-3.52L3.52,12L6,9.52V6h3.52L12,3.52L14.48,6H18v3.52L20.48,12L18,14.48z"}),a.a.createElement("path",{d:"M12,6c-3.31,0-6,2.69-6,6s2.69,6,6,6s6-2.69,6-6S15.31,6,12,6z M12,16c-2.21,0-4-1.79-4-4s1.79-4,4-4c2.21,0,4,1.79,4,4 S14.21,16,12,16z"}),a.a.createElement("circle",{cx:"12",cy:"12",r:"2.5"}))))}function K(e){return a.a.createElement(Q.a,e,a.a.createElement("g",{id:"Bounding_Boxes"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy_2"}),a.a.createElement("path",{fill:"none",d:"M0,0h24v24H0V0z"})),a.a.createElement("g",{id:"Duotone"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy_3"}),a.a.createElement("g",null,a.a.createElement("path",{opacity:"0.3",d:"M12,4c-4.42,0-8,3.58-8,8c0,4.42,3.58,8,8,8s8-3.58,8-8C20,7.58,16.42,4,12,4z M15.5,8 C16.33,8,17,8.67,17,9.5S16.33,11,15.5,11S14,10.33,14,9.5S14.67,8,15.5,8z M8.5,8C9.33,8,10,8.67,10,9.5S9.33,11,8.5,11 S7,10.33,7,9.5S7.67,8,8.5,8z M12,17.5c-2.33,0-4.32-1.45-5.12-3.5h1.67c0.7,1.19,1.97,2,3.45,2s2.76-0.81,3.45-2h1.67 C16.32,16.05,14.33,17.5,12,17.5z"}),a.a.createElement("circle",{cx:"15.5",cy:"9.5",r:"1.5"}),a.a.createElement("circle",{cx:"8.5",cy:"9.5",r:"1.5"}),a.a.createElement("path",{d:"M12,16c-1.48,0-2.75-0.81-3.45-2H6.88c0.8,2.05,2.79,3.5,5.12,3.5s4.32-1.45,5.12-3.5h-1.67C14.76,15.19,13.48,16,12,16z"}),a.a.createElement("path",{d:"M11.99,2C6.47,2,2,6.48,2,12c0,5.52,4.47,10,9.99,10C17.52,22,22,17.52,22,12C22,6.48,17.52,2,11.99,2z M12,20 c-4.42,0-8-3.58-8-8c0-4.42,3.58-8,8-8s8,3.58,8,8C20,16.42,16.42,20,12,20z"}))))}function X(e){return a.a.createElement(Q.a,e,a.a.createElement("g",{id:"Bounding_Boxes"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy_2"}),a.a.createElement("path",{fill:"none",d:"M0,0h24v24H0V0z"})),a.a.createElement("g",{id:"Duotone"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy"}),a.a.createElement("g",null,a.a.createElement("path",{opacity:"0.3",d:"M12,4c-4.42,0-8,3.58-8,8c0,4.42,3.58,8,8,8s8-3.58,8-8C20,7.58,16.42,4,12,4z M6.76,8.82l1.06-1.06 l1.06,1.06l1.06-1.06L11,8.82L9.94,9.88L11,10.94L9.94,12l-1.06-1.06L7.82,12l-1.06-1.06l1.06-1.06L6.76,8.82z M6.89,17 c0.8-2.04,2.78-3.5,5.11-3.5s4.31,1.46,5.11,3.5H6.89z M17.24,10.94L16.18,12l-1.06-1.06L14.06,12L13,10.94l1.06-1.06L13,8.82 l1.06-1.06l1.06,1.06l1.06-1.06l1.06,1.06l-1.06,1.06L17.24,10.94z"}),a.a.createElement("path",{d:"M12,13.5c-2.33,0-4.31,1.46-5.11,3.5h10.22C16.31,14.96,14.33,13.5,12,13.5z"}),a.a.createElement("polygon",{points:"7.82,12 8.88,10.94 9.94,12 11,10.94 9.94,9.88 11,8.82 9.94,7.76 8.88,8.82 7.82,7.76 6.76,8.82 7.82,9.88  6.76,10.94 \t\t"}),a.a.createElement("path",{d:"M11.99,2C6.47,2,2,6.47,2,12c0,5.53,4.47,10,9.99,10S22,17.53,22,12C22,6.47,17.52,2,11.99,2z M12,20c-4.42,0-8-3.58-8-8 c0-4.42,3.58-8,8-8s8,3.58,8,8C20,16.42,16.42,20,12,20z"}),a.a.createElement("polygon",{points:"16.18,7.76 15.12,8.82 14.06,7.76 13,8.82 14.06,9.88 13,10.94 14.06,12 15.12,10.94 16.18,12 17.24,10.94  16.18,9.88 17.24,8.82 \t\t"}))))}function Y(e){return a.a.createElement(Q.a,e,a.a.createElement("g",{id:"Bounding_Boxes"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy_2"}),a.a.createElement("path",{fill:"none",d:"M0,0h24v24H0V0z"})),a.a.createElement("g",{id:"Duotone"},a.a.createElement("g",{id:"ui_x5F_spec_x5F_header_copy"}),a.a.createElement("g",null,a.a.createElement("path",{opacity:"0.3",d:"M12,4c-4.42,0-8,3.58-8,8c0,4.42,3.58,8,8,8s8-3.58,8-8C20,7.58,16.42,4,12,4z M8.88,7.82L11,9.94L9.94,11 L8.88,9.94L7.82,11L6.76,9.94L8.88,7.82z M12,17.5c-2.33,0-4.31-1.46-5.11-3.5h10.22C16.31,16.04,14.33,17.5,12,17.5z M16.18,11 l-1.06-1.06L14.06,11L13,9.94l2.12-2.12l2.12,2.12L16.18,11z"}),a.a.createElement("polygon",{points:"8.88,9.94 9.94,11 11,9.94 8.88,7.82 6.76,9.94 7.82,11 \t\t"}),a.a.createElement("polygon",{points:"13,9.94 14.06,11 15.12,9.94 16.18,11 17.24,9.94 15.12,7.82 \t\t"}),a.a.createElement("path",{d:"M11.99,2C6.47,2,2,6.47,2,12c0,5.53,4.47,10,9.99,10S22,17.53,22,12C22,6.47,17.52,2,11.99,2z M12,20c-4.42,0-8-3.58-8-8 c0-4.42,3.58-8,8-8s8,3.58,8,8C20,16.42,16.42,20,12,20z"}),a.a.createElement("path",{d:"M12,17.5c2.33,0,4.31-1.46,5.11-3.5H6.89C7.69,16.04,9.67,17.5,12,17.5z"}))))}function Z(e){switch(e.gameStatus){case W.won:return a.a.createElement(Y,{className:e.classes.icon,style:{color:O.palette.state.good,fontSize:"48px"}});case W.lost:return a.a.createElement(X,{className:e.classes.icon,style:{color:O.palette.state.bad,fontSize:"48px"}});case W.initialised:case W.started:default:return a.a.createElement(K,{className:e.classes.icon,style:{color:O.palette.primary.dark,fontSize:"48px"}})}}var ee=function(e){var t=[],i=0,n="undefined"===typeof e.minefieldStyle?"100%":e.minefieldStyle.fontSize,s=!0,l=!1,r=void 0;try{for(var c,o=e.minefield[Symbol.iterator]();!(s=(c=o.next()).done);s=!0){var d=c.value,m="";if(d.revealed)switch(d.value){case 0:m="";break;case"*":m=a.a.createElement(q,{className:"".concat(e.classes.icon," ").concat(e.trippedMineId===d.id?"pulse":""),style:{fontSize:n,color:e.gameStatus===W.won?O.palette.state.good:O.palette.state.bad}});break;default:m=d.value}var u=void 0;switch(d.flag){case V:u=a.a.createElement($,{className:e.classes.icon,style:{fontSize:n}});break;case R:u="?";break;case D:default:u=""}t.push(a.a.createElement("div",{className:"scene",key:i},a.a.createElement("div",{className:"card ".concat(d.revealed?"reveal":""," delay-").concat(d.delay),onClick:e.handleClick.bind(void 0,i),onContextMenu:e.handleClick.bind(void 0,i)},a.a.createElement("div",{className:"card-face front"},u),a.a.createElement("div",{className:"card-face back"},m)))),i++}}catch(h){l=!0,r=h}finally{try{s||null==o.return||o.return()}finally{if(l)throw r}}return t},te=function(e){function t(e){var i;return Object(c.a)(this,t),(i=Object(d.a)(this,Object(m.a)(t).call(this,e))).componentDidMount=function(){var e,t,n=i.getMinefieldDimensions(i.state.difficulty);e=n.columns,t=n.rows,i.setState({columns:e,rows:t}),i.setColumnsInCSSGrid(e),i.model.columns=e,i.model.rows=t,i.initialiseMinefield(e*t,i.state.mines,e)},i.componentWillUnmount=function(){i.timer.current.resetTimer()},i.getMinefieldDimensions=function(e){var t=window.innerWidth,n=8;return"easy"!==e&&(n=i.getNumberOfColumnsForMinefield(t)),{columns:n,rows:i.getNumberOfRowsForMinefield(n,e)}},i.getNumberOfColumnsForMinefield=function(e){return e<=480?8:e<=768?12:e<=1024?18:24},i.getNumberOfRowsForMinefield=function(e,t){switch(t){default:case"easy":return 72/e;case"medium":return 288/e;case"hard":return 504/e}},i.handleChangeDifficulty=function(e){var t=e.target.value;i.resetGame(t)},i.initialiseMinefield=function(e,t,n){i.model.initialiseMinefield(e,t,n),i.setState(function(e,t){return{minefield:i.model.minefield.slice()}})},i.resetGame=function(e){var t,n,a;switch(i.timer.current.resetTimer(),"string"!==typeof e&&(e=i.state.difficulty),e){default:case"easy":n=10;break;case"medium":n=40;break;case"hard":n=100}var s=i.getMinefieldDimensions(e);t=s.columns,a=s.rows,i.setColumnsInCSSGrid(t),i.setState({columns:t,difficulty:e,mines:n,minesToBeFound:n,rows:a}),i.model.columns=t,i.model.rows=a,i.initialiseMinefield(a*t,n,t)},i.setColumnsInCSSGrid=function(e){var t,n,a,s=window.innerHeight-64,l=window.innerWidth,c="calc(var(--content-width) / var(--columns) * ".concat(s>l?.5:.6,")");8===e?(n=s>l?"calc(96vmin)":"calc(80vmin - 64px)",a="1vh"):(n="96vw",a="1vw"),i.setState({minefieldStyle:(t={},Object(r.a)(t,"--columns",e),Object(r.a)(t,"--content-width",n),Object(r.a)(t,"--gutter",a),Object(r.a)(t,"fontSize",c),t)})},i.classes=e.classes,i.model=new J,i.state={difficulty:["easy","medium","hard"].includes(i.props.difficulty)?i.props.difficulty:"easy",minefield:i.model.minefield.slice(),mines:10,minesToBeFound:10,time:null},i.timer=a.a.createRef(),i.handleClick=i.handleClick.bind(Object(h.a)(Object(h.a)(i))),i}return Object(u.a)(t,e),Object(o.a)(t,[{key:"handleClick",value:function(e,t){if(t.preventDefault(),this.model.gameStatus===W.initialised||this.model.gameStatus===W.started)switch(t.type){case"click":if(this.model.gameStatus===W.initialised&&(this.timer.current.resetTimer(),this.timer.current.startTimer()),!this.model.isRevealed(e)){this.model.reveal(e);var i=this.state.minesToBeFound;this.model.gameStatus!==W.won&&this.model.gameStatus!==W.lost||(this.timer.current.stopTimer(),i=0),this.setState({minefield:this.model.minefield.slice(),minesToBeFound:i})}break;case"contextmenu":var n=this.state.minesToBeFound,a=this.model.cycleFlag(e);a===V&&n--,a===R&&n++,this.setState({minefield:this.model.minefield.slice(),minesToBeFound:n})}}},{key:"render",value:function(){return a.a.createElement(N.a,{theme:O},a.a.createElement(p.a,null),a.a.createElement("div",{className:"App"},a.a.createElement(M.a,{position:"fixed"},a.a.createElement(b.a,{disableGutters:!0,style:j},a.a.createElement(_.a,{className:this.classes.margin,color:"",value:this.state.difficulty,onChange:this.handleChangeDifficulty,inputProps:{name:"difficulty",id:"select-difficulty"}},a.a.createElement(L.a,{value:"easy"},"Easy"),a.a.createElement(L.a,{value:"medium"},"Medium"),a.a.createElement(L.a,{value:"hard"},"Hard")),a.a.createElement(k.a,{variant:"fab",color:"default","aria-label":"reset",onClick:this.resetGame},a.a.createElement(Z,{classes:this.classes,gameStatus:this.model.gameStatus})),a.a.createElement(y.a,{className:this.classes.margin,color:"inherit",variant:"subheading"},a.a.createElement(F.a,{className:this.classes.padding,badgeContent:this.state.minesToBeFound,color:"secondary"},"Mines")),a.a.createElement(H,{time:this.state.time,classes:this.classes,ref:this.timer}))),a.a.createElement("div",{className:"minefield ".concat(this.state.difficulty),style:this.state.minefieldStyle},a.a.createElement(ee,{classes:this.classes,minefield:this.state.minefield,minefieldStyle:this.state.minefieldStyle,handleClick:this.handleClick,trippedMineId:this.model.trippedMineId,gameStatus:this.model.gameStatus}))))}}]),t}(n.Component),ie=Object(f.withStyles)(function(e){return{margin:{margin:2*e.spacing.unit},padding:{padding:"0 ".concat(2*e.spacing.unit,"px")}}})(te),ne=Boolean("localhost"===window.location.hostname||"[::1]"===window.location.hostname||window.location.hostname.match(/^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/));function ae(e){navigator.serviceWorker.register(e).then(function(e){e.onupdatefound=function(){var t=e.installing;t.onstatechange=function(){"installed"===t.state&&(navigator.serviceWorker.controller?console.log("New content is available; please refresh."):console.log("Content is cached for offline use."))}}}).catch(function(e){console.error("Error during service worker registration:",e)})}l.a.render(a.a.createElement(ie,null),document.getElementById("root")),function(){if("serviceWorker"in navigator){if(new URL("/react-minesweeper",window.location).origin!==window.location.origin)return;window.addEventListener("load",function(){var e="".concat("/react-minesweeper","/service-worker.js");ne?(function(e){fetch(e).then(function(t){404===t.status||-1===t.headers.get("content-type").indexOf("javascript")?navigator.serviceWorker.ready.then(function(e){e.unregister().then(function(){window.location.reload()})}):ae(e)}).catch(function(){console.log("No internet connection found. App is running in offline mode.")})}(e),navigator.serviceWorker.ready.then(function(){console.log("This web app is being served cache-first by a service worker. To learn more, visit https://goo.gl/SC7cgQ")})):ae(e)})}}()},93:function(e,t,i){e.exports=i(237)},98:function(e,t,i){}},[[93,2,1]]]);
//# sourceMappingURL=main.01e00356.chunk.js.map