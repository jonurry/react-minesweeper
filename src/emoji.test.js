import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Emoji from './emoji.js';
import { GAME_STATUS } from './model.js';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

test('render won, happy and lost emoji', () => {
  const won = shallow(
    <Emoji gameStatus={GAME_STATUS.won} classes={{ icon: null }} />
  );
  expect(won).toMatchSnapshot();
});
