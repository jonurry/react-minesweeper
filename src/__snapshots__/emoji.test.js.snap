// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render won, happy and lost emoji 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Emoji
    classes={
      Object {
        "icon": null,
      }
    }
    gameStatus={2}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "className": null,
      "style": Object {
        "color": "#81c784",
        "fontSize": "48px",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "className": null,
        "style": Object {
          "color": "#81c784",
          "fontSize": "48px",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
